File: app/src/main/java/com/raju/javabaseproject/data/source/remote/utils/HttpClientBuilder.java
package com.raju.javabaseproject.data.source.remote.utils;

import android.content.Context;

import com.raju.javabaseproject.data.source.remote.utils.interceptors.CacheInterceptor;
import com.raju.javabaseproject.data.source.remote.utils.interceptors.ConnectivityInterceptor;
import com.raju.javabaseproject.data.source.remote.utils.interceptors.HeaderInterceptor;
import com.raju.javabaseproject.data.source.remote.utils.interceptors.OfflineInterceptor;

import java.io.File;
import java.io.IOException;
import java.util.concurrent.TimeUnit;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.X509TrustManager;

import okhttp3.Cache;
import okhttp3.Connection;
import okhttp3.CookieJar;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Response;
import timber.log.Timber;

public class HttpClientBuilder {

    private int timeoutConnect;
    private int timeoutRead;
    private int timeoutWrite;

    private String cacheName;
    private int cacheSizeMB;
    private boolean enableSocketLog;

    private SSLSocketFactory unsafeSSLSocketFactory;
    private X509TrustManager x509TrustManager;
    private HostnameVerifier hostnameVerifier;

    private Interceptor networkInterceptor, loggingInterceptor, wireFormatHeaderInterceptor;

    private CookieJar cookieJar;

    private final Context context;

    public HttpClientBuilder(Context context) {
        this.context = context;
    }

    public OkHttpClient build() {
        OkHttpClient.Builder okClientBuilder = new OkHttpClient.Builder();

        okClientBuilder.addInterceptor(new ConnectivityInterceptor(context));
        if (null != loggingInterceptor) {
            okClientBuilder.addInterceptor(loggingInterceptor);
        }

        final File baseDir = context.getCacheDir();
        if (baseDir != null) {
            final File cacheDir = new File(baseDir, cacheName);
            int cacheSize = cacheSizeMB * 1024 * 1024;
            okClientBuilder.cache(new Cache(cacheDir, cacheSize));
        }

        okClientBuilder.connectTimeout(this.timeoutConnect, TimeUnit.SECONDS);
        okClientBuilder.readTimeout(this.timeoutRead, TimeUnit.SECONDS);
        okClientBuilder.writeTimeout(this.timeoutWrite, TimeUnit.SECONDS);

        if (unsafeSSLSocketFactory != null) {
            okClientBuilder.sslSocketFactory(unsafeSSLSocketFactory, x509TrustManager);
            okClientBuilder.hostnameVerifier(hostnameVerifier);
        }

        if (null != networkInterceptor) {
            okClientBuilder.addNetworkInterceptor(networkInterceptor);
        }
        if (null != wireFormatHeaderInterceptor) {
            okClientBuilder.addInterceptor(wireFormatHeaderInterceptor);
        }

        if (null != cookieJar) {
            okClientBuilder.cookieJar(cookieJar);
        }

        if (enableSocketLog) {
            okClientBuilder.addNetworkInterceptor(new Interceptor() {
                @Override
                public Response intercept(Chain chain) throws IOException {
                    Connection conn = chain.connection();
                    return chain.proceed(chain.request());
                }
            });
        }

        okClientBuilder.addInterceptor(new HeaderInterceptor());
        okClientBuilder.addInterceptor(new OfflineInterceptor(context));
        okClientBuilder.addNetworkInterceptor(new CacheInterceptor());
        okClientBuilder.cache(provideCache());
        return okClientBuilder.build();
    }

    public HttpClientBuilder setCache(String cacheName, int cacheSizeMB) {
        this.cacheName = cacheName;
        this.cacheSizeMB = cacheSizeMB;
        return this;
    }

    public HttpClientBuilder setTimeouts(int timeout) {
        return setConnectTimeout(timeout)
                .setReadTimeout(timeout)
                .setWriteTimeout(timeout);
    }

    public HttpClientBuilder setConnectTimeout(int timeout) {
        this.timeoutConnect = timeout;
        return this;
    }

    public HttpClientBuilder setReadTimeout(int timeout) {
        this.timeoutRead = timeout;
        return this;
    }

    public HttpClientBuilder setWriteTimeout(int timeout) {
        this.timeoutWrite = timeout;
        return this;
    }

    public HttpClientBuilder setNetworkInterceptor(Interceptor interceptor) {
        this.networkInterceptor = interceptor;
        return this;
    }

    public HttpClientBuilder setWireFormatHeaderInterceptor(Interceptor wireFormatHeaderInterceptor) {
        this.wireFormatHeaderInterceptor = wireFormatHeaderInterceptor;
        return this;
    }

    public HttpClientBuilder setLoggingInterceptor(Interceptor interceptor) {
        this.loggingInterceptor = interceptor;
        return this;
    }

    public HttpClientBuilder setCookieJar(CookieJar cookieJar) {
        this.cookieJar = cookieJar;
        return this;
    }

    public HttpClientBuilder enableSocketLog(boolean enable) {
        this.enableSocketLog = enable;
        return this;
    }

    public HttpClientBuilder setUnsafeSSLSocketFactory(SSLSocketFactory factory,
                                                       X509TrustManager trustManager,
                                                       HostnameVerifier hostnameVerifier) {
        this.unsafeSSLSocketFactory = factory;
        this.x509TrustManager = trustManager;
        this.hostnameVerifier = hostnameVerifier;
        return this;
    }

    private Cache provideCache() {
        Cache cache = null;
        try {
            cache = new Cache(new File(context.getCacheDir(), "http-cache"),
                    10 * 1024 * 1024); // 10 MB
        } catch (Exception e) {
            Timber.d("Could not create Cache!");
        }
        return cache;
    }
}

File: app/src/main/java/com/raju/javabaseproject/data/source/remote/http/HttpClientBuilder.kt
package com.raju.javabaseproject.data.source.remote.http

import android.content.Context

import java.io.File
import java.util.concurrent.TimeUnit

import javax.net.ssl.HostnameVerifier
import javax.net.ssl.SSLSocketFactory
import javax.net.ssl.X509TrustManager

import okhttp3.Cache
import okhttp3.CookieJar
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import timber.log.Timber

class HttpClientBuilder(private val context: Context) {

    private var timeoutConnect: Int = 0
    private var timeoutRead: Int = 0
    private var timeoutWrite: Int = 0

    private var cacheName: String? = null
    private var cacheSizeMB: Int = 0
    private var enableSocketLog: Boolean = false

    private var unsafeSSLSocketFactory: SSLSocketFactory? = null
    private var x509TrustManager: X509TrustManager? = null
    private var hostnameVerifier: HostnameVerifier? = null

    private var networkInterceptor: Interceptor? = null
    private var loggingInterceptor: Interceptor? = null
    private var wireFormatHeaderInterceptor: Interceptor? = null

    private var cookieJar: CookieJar? = null

    fun build(): OkHttpClient {
        val okClientBuilder = OkHttpClient.Builder()

        okClientBuilder.addInterceptor(ConnectivityInterceptor(context))
        if (null != loggingInterceptor) {
            okClientBuilder.addInterceptor(loggingInterceptor!!)
        }

        val baseDir = context.cacheDir
        if (baseDir != null) {
            val cacheDir = File(baseDir, cacheName!!)
            val cacheSize = cacheSizeMB * 1024 * 1024
            okClientBuilder.cache(Cache(cacheDir, cacheSize.toLong()))
        }

        okClientBuilder.connectTimeout(this.timeoutConnect.toLong(), TimeUnit.SECONDS)
        okClientBuilder.readTimeout(this.timeoutRead.toLong(), TimeUnit.SECONDS)
        okClientBuilder.writeTimeout(this.timeoutWrite.toLong(), TimeUnit.SECONDS)

        if (unsafeSSLSocketFactory != null) {
            okClientBuilder.sslSocketFactory(unsafeSSLSocketFactory!!, x509TrustManager!!)
            okClientBuilder.hostnameVerifier(hostnameVerifier!!)
        }

        if (null != networkInterceptor) {
            okClientBuilder.addNetworkInterceptor(networkInterceptor!!)
        }
        if (null != wireFormatHeaderInterceptor) {
            okClientBuilder.addInterceptor(wireFormatHeaderInterceptor!!)
        }

        if (null != cookieJar) {
            okClientBuilder.cookieJar(cookieJar!!)
        }

        /*if (enableSocketLog) {
            okClientBuilder.addNetworkInterceptor(new Interceptor() {
                @Override
                public Response intercept(Chain chain) throws IOException {
                    Connection conn = chain.connection();
                    return chain.proceed(chain.request());
                }
            });
        }*/

        okClientBuilder.addInterceptor(OfflineInterceptor(context))
        //okClientBuilder.addNetworkInterceptor(new CacheInterceptor());
        okClientBuilder.cache(provideCache())
        return okClientBuilder.build()
    }

    fun setCache(cacheName: String, cacheSizeMB: Int): HttpClientBuilder {
        this.cacheName = cacheName
        this.cacheSizeMB = cacheSizeMB
        return this
    }

    fun setTimeouts(timeout: Int): HttpClientBuilder {
        return setConnectTimeout(timeout)
                .setReadTimeout(timeout)
                .setWriteTimeout(timeout)
    }

    fun setConnectTimeout(timeout: Int): HttpClientBuilder {
        this.timeoutConnect = timeout
        return this
    }

    fun setReadTimeout(timeout: Int): HttpClientBuilder {
        this.timeoutRead = timeout
        return this
    }

    fun setWriteTimeout(timeout: Int): HttpClientBuilder {
        this.timeoutWrite = timeout
        return this
    }

    fun setNetworkInterceptor(interceptor: Interceptor): HttpClientBuilder {
        this.networkInterceptor = interceptor
        return this
    }

    fun setWireFormatHeaderInterceptor(wireFormatHeaderInterceptor: Interceptor): HttpClientBuilder {
        this.wireFormatHeaderInterceptor = wireFormatHeaderInterceptor
        return this
    }

    fun setLoggingInterceptor(interceptor: Interceptor): HttpClientBuilder {
        this.loggingInterceptor = interceptor
        return this
    }

    fun setCookieJar(cookieJar: CookieJar): HttpClientBuilder {
        this.cookieJar = cookieJar
        return this
    }

    fun enableSocketLog(enable: Boolean): HttpClientBuilder {
        this.enableSocketLog = enable
        return this
    }

    fun setUnsafeSSLSocketFactory(factory: SSLSocketFactory,
                                  trustManager: X509TrustManager,
                                  hostnameVerifier: HostnameVerifier): HttpClientBuilder {
        this.unsafeSSLSocketFactory = factory
        this.x509TrustManager = trustManager
        this.hostnameVerifier = hostnameVerifier
        return this
    }

    private fun provideCache(): Cache? {
        var cache: Cache? = null
        try {
            cache = Cache(File(context.cacheDir, "http-cache"),
                    (10 * 1024 * 1024).toLong()) // 10 MB
        } catch (e: Exception) {
            Timber.d("Could not create Cache!")
        }

        return cache
    }
}
